contrib/xor-blocks/src/main.rs

```diff
@@ -0,0 +1,130 @@
+use std::{
+    env, fs,
+    io::{self, Read},
+    path::Path,
+    time::{Duration, Instant},
+};
+
+const MAGIC: [u8; 4] = [0xf9, 0xbe, 0xb4, 0xd9];
+
+fn main() -> Result<(), io::Error> {
+    let start = Instant::now();
+
+    let blocks_path = if env::args().len() > 1 {
+        env::args().nth(1).expect("the arg to exist").into()
+    } else {
+        #[allow(deprecated)]
+        env::home_dir()
+            .expect("to have a home directory")
+            .join(match env::consts::OS {
+                "macos" => "Library/Application Support/Bitcoin",
+                "windows" => "AppData\\Local\\Bitcoin",
+                "linux" => ".bitcoin",
+                _ => {
+                    println!("Unknown OS");
+                    return Ok(());
+                }
+            })
+            .join("blocks")
+    };
+
+    let paths = fs::read_dir(&blocks_path)?
+        .map(|res| res.map(|e| e.path()))
+        .collect::<Result<Vec<_>, io::Error>>()?;
+
+    let xor_path: std::path::PathBuf = blocks_path.join("xor.dat");
+    if !fs::exists(&xor_path)? {
+        println!("No xor.dat file. Make sure you are running Bitcoin Core v28 or higher.");
+        return Ok(());
+    }
+
+    println!("Xor'ing blocks dir. Do not start bitcoind until finished!");
+
+    let key: [u8; 8] = fs::read(&xor_path)?
+        .try_into()
+        .expect("xor.dat to be 8 bytes");
+    if key[0..4] == [0u8; 4] && key[4..] != [0u8; 4] {
+        println!("This script doesn't work with a non-zero key with 4 bytes of leading zeros");
+        return Ok(());
+    }
+    let key = if key == [0u8; 8] {
+        loop {
+            let key: [u8; 8] = rand::random();
+            // Don't use keys with 4 bytes of leading zeros
+            // They won't let us detect the first 4 bytes of magic in the files
```

Comment:
Why though? After "unxoring" the magic bytes just will be the same. Anyway, might still be a good idea to mask it better.

Reply:
we would reapply the xor on restart in that case, since we wouldn't detect that we've processed it already.

Alternatively we could keep a counter of processed blocks - modify the `xor.dat` to make sure we can't start `bitcoind` (or @LarryRuane's lock would probably also do the same) and add a height of how many blocks we've processed so far (though that would make parallelization more difficult). When we're done we'd restore the `xor.dat` file (which we have to modify anyway if we're xor-ing)

Reply:
Oh, I see, that's clever! But then yes, this code tries to be atomic but it isn't because it doesn't call `sync_data`.